File: Difficulty 4.8/Prime Sieve/PrimeSieve.java
import java.util.Scanner;
import java.util.BitSet;

public class PrimeSieve {
    public static void main (String [] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int num = sc.nextInt();

        // By default the value will be set to false
        BitSet prime = new BitSet(n + 1);

        for (int i = 2; i * i <= n; i++) {
            if (prime.get(i) == false) {
                for (int j = i * i; j <= n; j += i) {
                    prime.set(j);
                }
            }
        }

        int counter = 0;
        for (int i = 2; i <= n; i++) {
            if (prime.get(i) == false) {
                counter++;
            }
        }

        System.out.println(counter);

        for (int i = 0; i < num; i++) {
            int currNum = sc.nextInt();
            if (prime.get(currNum) || currNum == 1) {
                System.out.println(0);
            } else {
                System.out.println(1);
            }
        }
    }
}

File: Difficulty 4.8/Prime Sieve/PrimeSieve.py
line = str(input()).split(" ")
n = int(line[0])
num = int(line[1])

prime = [True for i in range(n + 1)]
p = 2

while (p * p <= n):
    if (prime[p] == True):
        # The current number is a prime
        for i in range(p * p, n + 1, p):
            prime[i] = False
    p += 1

counter = 0
for p in range(2, n + 1):
    if prime[p]:
        counter += 1

print(counter)

for x in range(num):
    currNum = int(input())
    if prime[currNum] and currNum != 1:
        print(1)
    else:
        print(0)
